src/parser/jobs/dnc/modules/Esprit.tsx

```diff
@@ -0,0 +1,189 @@
+	[ACTIONS.SINGLE_TECHNICAL_FINISH.id]: 3,
+	[ACTIONS.DOUBLE_TECHNICAL_FINISH.id]: 3,
+	[ACTIONS.TRIPLE_TECHNICAL_FINISH.id]: 3,
+	[ACTIONS.QUADRUPLE_TECHNICAL_FINISH.id]: 3,
+}
+
+const STANDARD_FINISHES = [
+	ACTIONS.STANDARD_FINISH.id,
+	ACTIONS.SINGLE_STANDARD_FINISH.id,
+	ACTIONS.DOUBLE_STANDARD_FINISH.id,
+]
+
+const TECHNICAL_FINISHES = [
+	ACTIONS.TECHNICAL_FINISH.id,
+	ACTIONS.SINGLE_TECHNICAL_FINISH.id,
+	ACTIONS.DOUBLE_TECHNICAL_FINISH.id,
+	ACTIONS.TRIPLE_TECHNICAL_FINISH.id,
+	ACTIONS.QUADRUPLE_TECHNICAL_FINISH.id,
+]
+
+const FINISHES = [
+	...STANDARD_FINISHES,
+	...TECHNICAL_FINISHES,
+]
+
+const ESPRIT_GENERATION_AMOUNT = 10
+
+const TICK_FREQUENCY = 3000
+const MAX_IMPROV_TICKS = 5
+
+const ESPRIT_RATE_SELF = 0.25
+const ESPRIT_RATE_PARTY = 0.2
+
+const MAX_ESPRIT = 100
+const SABER_DANCE_COST = 50
+
+export default class EspritGauge extends Module {
+	static handle = 'espritgauge'
+	static title = t('dnc.esprit-gauge.title')`Esprit Gauge`
+
+	@dependency private combatants!: Combatants
+	@dependency private suggestions!: Suggestions
+
+	private potentialOvercap = 0
+	private espritConsumed = 0
+	private avgGenerated = 0
+	private history: any[] = []
+	private currentEsprit = 0
+	private improvisationStart = 0
+
+	protected init() {
+		this.addHook('damage', {by: 'player'}, this.onDamage)
+		this.addHook('cast', {by: 'player', abilityId: ACTIONS.SABER_DANCE.id}, this.onConsumeEsprit)
+		this.addHook('applybuff', {by: 'player', abilityId: STATUSES.IMPROVISATION.id}, this.startImprov)
+		this.addHook('removebuff', {by: 'player', abilityId: STATUSES.IMPROVISATION.id}, this.endImprov)
+		this.addHook('death', {to: 'player'}, this.onDeath)
+		this.addHook('complete', this.onComplete)
+	}
+	private onDamage(event: DamageEvent) {
+		if (!ESPRIT_GENERATION_MULTIPLIERS[event.ability.guid] || event.amount === 0) {
+			return
+		}
+		let generatedAmt = 0
+		if (this.combatants.selected.hasStatus(STATUSES.TECHNICAL_FINISH.id)) {
+			generatedAmt += ESPRIT_GENERATION_MULTIPLIERS[event.ability.guid] * ESPRIT_GENERATION_AMOUNT * ESPRIT_RATE_PARTY * (Object.keys(this.combatants.getEntities()).length-1)
+			// Finishes aren't a weaponskill, so they don't generate esprit
+			if (!FINISHES[event.ability.guid]) {
+				generatedAmt += ESPRIT_GENERATION_MULTIPLIERS[event.ability.guid] * ESPRIT_GENERATION_AMOUNT * ESPRIT_RATE_SELF
+			}
+		} else if (this.combatants.selected.hasStatus(STATUSES.ESPRIT.id)) {
+			generatedAmt += ESPRIT_GENERATION_MULTIPLIERS[event.ability.guid] * ESPRIT_GENERATION_AMOUNT * ESPRIT_RATE_SELF
+			if (this.combatants.selected.hasStatus(STATUSES.CLOSED_POSITION.id)) {
+				generatedAmt += ESPRIT_GENERATION_MULTIPLIERS[event.ability.guid] * ESPRIT_GENERATION_AMOUNT * ESPRIT_RATE_PARTY
+			}
+		}
+		this.avgGenerated += generatedAmt
+		if (generatedAmt > 0) {
+			this.setEsprit(this.currentEsprit + generatedAmt)
+		}
+	}
+	private onConsumeEsprit() {
+		this.espritConsumed++
+		if (this.currentEsprit < SABER_DANCE_COST) {
```

Comment:
This is really close to kenki honestly - might be able to shore up the logic with some of that back-tracking magic. Then again maybe there's _too_ much unkown to esprit for it to be of any use.